### DISCUSSION TITLE
Methodology to calculate raw Metrics numbers

### DISCUSSION THREAD

**Discussion by schnuerle** [openmobilityfoundation/curb-data-specification#27] [Data Specification]
In order to get to Metrics, we need to define and specify the limits of measuring curb activity and usage for differing data sources and scenarios. 

After events come in from the Events API and are stored, how do those then get reconciled into curb activity like event counts, counts over times, dwell time, duration, and usage percents?

Events can be from on/off sensors, or camera devices, in person recording, fleet data feeds, etc, each collecting different data or not knowing other data (eg, vehicle type, vehicle length, unused curb linear feet). How do you count events, aggregate events, and overlapping events from different sensors or sources?

Is it ok to caveat the Metrics as sampling, off model, or representative data, vs being a 100% accurate view of the curb at all times?  

After [defining use cases](https://github.com/openmobilityfoundation/curb-data-specification/discussions/28) we can ensure this methodology helps meet them.

Current Metrics could include (see [draft document](https://docs.google.com/document/d/1UgD2fHtVyIMwNG-qXJRv-gcY7nCdYxokbMsLTs32Em4/edit#heading=h.gsqrmtuugncm)):
1. Total Events
2. Turnover
3. Average Dwell Time
4. Occupancy Percent
5. Curb Productivity Index

**Comment by jacobmalleau** [comment]
Curious to know how others have typically calculated turnover. I have often seen it divided by the number of spaces for the zone being looked at. See here: 
https://www.civil.iitb.ac.in/tvm/1100_LnTse/535_lnTse/plain/plain.html

**Comment by schnuerle** [comment]
Regarding turnover, from the link above:

```
It is the ratio of number of vehicles parked in a duration to the number of parking bays available.
    parking turnover= parking volume/No. of bays available
This can be expressed as number of vehicles per bay per time duration.
```
Which is saying `vehicles per space/time`

In our draft we have a more simple 

```
    count[events]/hour 
Used together with Average Dwell Time by cities to understand how long vehicles are parked at the curb. 
When evaluated alongside a vehicle type breakdown, cities can see if the curb space is being used as intended 
and design better rules for compliance.
```

For CDS we have events, which could be a space for one vehicle, or the use of a curb area that is unmarked. So I feel these two boil down to the same thing. Though we are specifying the time unit of hour.

**Comment by bhamlinSDOT** [comment]
I _think_ we are leaning towards caveating that Metrics (and Events for that matter) are a sampling of data and that they are not a 100% inclusive depiction of all the curbside events. Parking studies, enforcement records, sensor data, cameras, etc. will all exist in separate "event" silos and I think as long as the required information is provided, then the Metrics can be calculated using our methodology. For example, if you want to calculate Metrics using enforcement data, you won't be able to produce all of the proposed Metrics because you won't have all of the required information. (maybe you could do total events for a particular time period but enforcement and parking studies are often  just a snap shot in time and not check in/check out) Using your sensor data though, you will be able to calculate all of the Metrics. but it will be challenging to combine the two together.

That said, if we have common units of measurements, an end user could pull metrics from the 2 different "silo's" and do further analysis beyond the scope of V1 of CDS.

**Comment by schnuerle** [comment]
Added a new "Representative Sample Data" section in Metrics to start putting guardrails on how the methodology can be calculated across sources.

https://github.com/openmobilityfoundation/curb-data-specification/blob/feature-release-work-1/metrics/README.md#representative-sample-data

**Comment by schnuerle** [comment]
We still need some work on now to take Events from the different source types and get them into some of the metrics, like Average Dwell Time and therefore Occupancy Percent. 

https://github.com/openmobilityfoundation/curb-data-specification/tree/feature-release-work-1/metrics#methodology

**Reply by schnuerle** [reply]
This is now sussed out with the reworking of /activity to /session. All of these can be calculated now using /session data.

Added a sentence to point this out here:
https://github.com/openmobilityfoundation/curb-data-specification/blob/feature-release-work-1/metrics/README.md#methodology

**Comment by schnuerle** [comment]
We may want to consider a third endpoint (or maybe as a replacement for /metrics/activities) called /metrics/sessions (or possibly /metrics/transactions, though there is likely no monetary transaction taking place) that is a conceptual level between /activities and /aggregates, that uses /events/event data to calculate a vehicle's "session" at the curb, i.e. its curb usage.  Would include only start and end times of a defined session of a curb user.

We need something like this to determine occupancy in curb places for our Metrics.

**Reply by schnuerle** [reply]
/activities is good as an easy thing to implement (simple subset of Events) with privacy up-front for researchers or the public. Is it worth the additional complexity it adds to the spec?

For /sessions (could be parking sessions ) we could use /activities and split up 'event_time' into 'session_start' and 'session_end' and use park_start and park_end as the Event types for this. Duration can be calculated from these new fields. If there is no park start/end then don't send anything, if there is only one of those then is useful to send that? You now something happened there.

**Reply by bhamlinSDOT** [reply]
It seems like /sessions in some capacity a stand along object. /activities provides the most granular data of each individual event (check-in (1 event) check-out (1 event), /aggregate provides summarized Metric calculations, and /sessions is a way to see rolled up individual parking sessions at a particular curb zone which is required to calculate certain Metrics.

I don't know if this need to be it's own endpoint or if what we need can be accomplished by including sessions into the Events API and then calculated for Metrics?

**Reply by schnuerle** [reply]
I've changed /activity to /session across Metrics. It contains all the same data that /activity had, but it's organized into the concept of a session, with each row of data representing a parking session start/end or an area session enter/exit. 

https://github.com/openmobilityfoundation/curb-data-specification/blob/feature-release-work-1/metrics/README.md#session

**Reply by jacobmalleau** [reply]
I think we are still missing a component here. The structure is set up such that all sessions are made up of events (typically 2). Although this is technically true, some times we may get data from companies as sessions (e.g. payment companies providing payment data with a start and end of session, i.e. 1 session composing of 2 events).
Therefore I think we should have an optional sessionID in the Metrics API to denote this type of data coming in. And Also have the sessionID as an optional attribute in the Events API in case the event was derived from a session. 

Thoughts?

**Reply by schnuerle** [reply]
I had added session_id [here](https://github.com/openmobilityfoundation/curb-data-specification/discussions/27#discussioncomment-1683845) after a WG action item, then removed it again after a later WG/WGSC discussion. 

If data comes from a company as a session already, can't it just be added as 2 separate events to CDS through Events? That still keeps it as 2 Events that can be connected in Metrics, without the need for a session_id.

Are you thinking that there is some extra value in having an explicit session_id in Events to know for sure these events are connected @jacobmalleau ?

**Reply by jacobmalleau** [reply]
I guess it comes down to how we want to define a unique session. Is it good enough to define it by its start and end events? Do you envision every set of events to be grouped into sessions, and then further grouped into aggregate for metrics? I see having the session_id in Events (as optional) provides an easy way to go between events <-> sessions <-> metrics. Without the session_id we lose an arrow (events <- sessions <-> metrics). I imagine a lot of payment companies will provide their source data as sessions (typical for transaction data) so defining a unique session_id to link back to this is important.

Another application where session_id could be helpful:
Working with real-time data for Visualization: I think platforms/agences would typically want to deal with the Events API when dealing with real time data and displaying that for users. If so, it becomes more involved to calculate demand for a given space if Session IDs are not provided. The system would have to pull all Events and sum all park starts and subtract all park ends to get overall occupancy. If session_ids are provided, a user just have to simply check how many sessions have a park_start and no park_end occurring at the given time. (Main counter point to this is that you could just query Sessions at this point, but it gives the platform flexibility).

**Reply by schnuerle** [reply]
If we think about how Events will be used, we want them to come in as they happen in CDS.  So when a vehicle pulls up and stops to occupy a curb space, a park_start Event would be sent.  Then later when it leaves a park_end Event is sent. 

This allows cities to track real-time occupancy (if the can and want to). We wouldn't want data to come through only when a session is over.  

But as long as a session_id is created at the start, and the same ID is used later at the end, then that would be ok.

But we wouldn't want to require session_id of course, and I image most data sources might not be able to provide a session_id at the start and end of a session?  So there would still have to be manual work done like you mention when there is no session_id.

Having said all that, I'm inclined to add it as optional with some guidance about ensuring it is provided and consistent with park_start and park_end events.

**Reply by jacobmalleau** [reply]
Agree with having it optional and ensuring it is consistent with park_start and park_end events!
My one counter point to the instance you gave above is that if you are dealing with transaction/app payment data, you actually don't know when the user has left a curb space. They may pay for an hour (and in turn you would get the park_start time, duration, and associated park_end time at the start of the session) but you wouldn't know if they stayed the full hour, or left before or after their time is up.

**Reply by schnuerle** [reply]
I would say in that case you would 1) not use session_id and only record a park_start or 2) use session_id and set the park_end to the maximum time and if you can set park_end to when a new vehicle is using the space and they begin a park_start. 

I think session_id should only be used when you know or can guess at the park start and end values, otherwise don't use session_id.

It makes me wonder though about adding a field like '**event_expected_duration**' so that on park_start events a company can specify the expected duration of a stay, eg 1 hour for a meter or a fleet can estimate a 15 min stay and send that. It seems like useful info to know, where you could look at historically, or when you don't get a park end event, or if you are supporting realtime availability you could publish expected leaving info.

**Reply by schnuerle** [reply]
I'll try and add session_id by tomorrow's meeting so we can discuss at WG. Any other thoughts on this or the duration field?

**Reply by schnuerle** [reply]
I've added `event_session_id` to the [Curbs Event](https://github.com/openmobilityfoundation/curb-data-specification/tree/feature-release-work-1/events#curb-event) and [Metrics Session](https://github.com/openmobilityfoundation/curb-data-specification/blob/feature-release-work-1/metrics/README.md#session) where applicable, with guidance. Please review and let me know if you have suggested changes.

For the `event_expected_duration` we may not add this based on the responses so far from the Working Group - it doesn't seem to be too valuable at this time.

**Comment by schnuerle** [comment]
Added an optional session_id that can be provided connect multiple event records into a cohesive session later for Metrics API endpoints. 

https://github.com/openmobilityfoundation/curb-data-specification/blob/feature-release-work-1/events/README.md#curb-event

**Reply by schnuerle** [reply]
Ok removed this in favor of calculating it later in the Metrics /session area.

**Comment by schnuerle** [comment]
I wonder if in the 'events' as defined in the Metrics Methodology really means 'session' and the wording is borrowed from the source of these metrics, and not aligning to the CDS language used.

Eg,
Total Events
count[events] for a specific time period
Name: total_events

should really be:

Total Sessions
count[sessions] for a specific time period
Name: total_sessions

Eliot Mueting may be able to answer this.

**Reply by ghost** [reply]
Yes, so by event we mean parking session.

**Reply by schnuerle** [reply]
Thank you @jean-populus. 

I've renamed total_events to total_sessions across Metrics and added a quick definition of what a session is: 
https://github.com/openmobilityfoundation/curb-data-specification/blob/feature-release-work-1/metrics/README.md#total-sessions
